### ISSUE TITLE
Feature: Using custom ConnectorEnvironment

### ISSUE THREAD

**Issue by maljac** [OCA/connector#107]
**Scenario**
When using the connector framework you have to use some API wrapper for your external system. This API wrapper maybe stateless or not. If your API wrapper is statefull, how do you pass the api around without creating the API instance and connecting each time again (which is costly)?

E.g. for reading the record, for reading children, for reading translation etc....

**Workaround**

In order to reuse the API instance, I wrote a custom ConnectorEnvironment which gets the API instance at instantiation:

```
class APIConnectorEnvironment(ConnectorEnvironment):
    def __init__(self, backend_record, session, model_name, api=None):
        super(APIConnectorEnvironment, self).__init__(
            backend_record, session, model_name)
        self.api = api

def get_environment(...):
    ...
    env = APIConnectorEnvironment(...., api=api)
    ....
    return env
```

Unfortunately this does not work all the time. In some situations the framework creates a new ConnectorEnvironment and it naturally does not use my custom ConnectorEnvironment class because of the `unit_for` method in the `ConnectorUnit`:

```
def unit_for(self, connector_unit_class, model=None):
        ...
        if model is None or model == self.model._name:
            env = self.connector_env
        else:
            env = ConnectorEnvironment(self.backend_record,
                                       self.session,
                                       model)
        return env.get_connector_unit(connector_unit_class)
```

So, in order to get my custom ConnectorEnvironment working I have to 

1) create a Mixin which must be used for all classes that inherit from `ConnectorUnit` (Exporter, Importer, Deleter, Synchronizer, Mapper, ...)
2) overwrite the `unit_for` method and _replace_ it in the mixin

**Feature Request**

For case (2) it would be nicer to have a hook method in `ConnectorUnit` like:

```
def unit_for(...):
        if model is None or model == self.model._name:
            env = self.connector_env
        else:
            env = self._create_connector_environment(model)
        return env.get_connector_unit(connector_unit_class)

    def _create_connector_environment(self, model):
        return ConnectorEnvironment(self.backend_record, self.session, model)
```

What solutions exist for (1) except for monkey patching ? Did I overlook something or should this be a supported feature: Interchangeable ConnectorEnvironment classes?

**Comment by guewen** [comment]
That's something I already thought about, but never had the use case so never took time to work on that. Definitely :+1: for implementing that.

I think that the hook in `ConnectorUnit` would not be ideal because we would still need to override it.

What about:
1. Create a new instance from the same class than the current env
2. Have an attribute or a way to declare which attributes needs to be propagated from env to env (maybe we need to make some research on the most elegant way to achieve that)
Which could look a bit like:

``` python

class ConnectorEnvironment(object):
    _propagate_args = ['backend_record', 'session', 'model_name']
    _propagate_kwargs = []

[...]

class APIConnectorEnvironment(ConnectorEnvironment):

    _propagate_kwargs = ['api']

    def __init__(self, backend_record, session, model_name, api=None):
        super(APIConnectorEnvironment, self).__init__(
            backend_record, session, model_name)
        self.api = api

[...]

    def unit_for(...):
        if model is None or model == self.model._name:
            env = self.connector_env
        else:
            env_class = self.connector_env.__class__
            args = env_class._propagate_args
            kwargs = {key: key for key in env_class._propagate_kwargs)
            env = env_class(*args, **kwargs)
        return env.get_connector_unit(connector_unit_class)
```

Would it work for you?

**Comment by maljac** [comment]
Looks good!

It seems that you only propagate the name of the attributes, not the values of the attributes. 

I see another problem with `_propagage_args`: Since the method `unit_for` expects the kwarg `model` which might be passed to the environment class (or not), depending on the `_propagate_args` list. How would you distinct between attributes of the `ConnectorUnit` and the kwarg `model`?

``` python

class ConnectorEnvironment(object):
    _propagate_args = ['backend_record', 'session', 'model_name']
    _propagate_kwargs = []

[...]

class APIConnectorEnvironment(ConnectorEnvironment):

    _propagate_kwargs = ['api']

    def __init__(self, backend_record, session, model_name, api=None):
        super(APIConnectorEnvironment, self).__init__(
            backend_record, session, model_name)
        self.api = api

[...]

def unit_for(self, connector_unit_class, model=None):
     if model is None or model == self.model._name:
            env = self.connector_env
        else:
            env_class = self.connector_env.__class__

            args = (getattr(self, arg) for arg in env_class._propagate_args
            # self.backend_record works
            # self.session works to
            # self.model_name name does not work / even self.model._name would be wrong

            # Solution (1) Instead of args = (getattr(..)) ...
            args = []
            _propagate_args = list(env_class._propagate_args) # !!!

            if 'model_name' in _propagate_args:
                _propagate_args.pop(_propagate_args.index('model_name'))
                args = [model]
            args = [getattr(self, arg) for arg in _propagate_args] + args
            kwargs = {key: getattr(self, key) for key in env_class._propagate_kwargs}

            env = env_class(*args, **kwargs)

            # Solution (2) Only use _propage_kwargs and keep the interface for positional args
            kwargs = {key: getattr(self, key) for key in env_class._propagate_kwargs}
            env = env_class(self.backend_record, self.session, model, **kwargs)

        return env.get_connector_unit(connector_unit_class)
[...]
```

_Solution 1_: Might be more "complete", but since the interface to `ConnectorEnvironment` might always need the three positional args, solution (2) could be sufficient.

_Solution 2_: Seems sufficient

Edit: If you want I can provide the PR with the code

**Comment by guewen** [comment]
> It seems that you only propagate the name of the attributes, not the values of the attributes.

Good catch ;-) Forgot the `getattr(self.connector_env, key)`.

Yes, solution 2 is fine. We can say that `args` are part of the core interface and are imposed and `kwargs` are open for extension, sounds good and cleaner than `1`.

I will be glad if you can provide the PR (even more if you include a test)  :-)

**Comment by lmignon** [comment]
I agree with @guewen `args`are part of the core api and I'm also in favor of the cleaner solution 2. @maljac what you propose is a good improvement.

**Comment by guewen** [comment]
Thinking further, this propagation of `kwargs` and creation of a new instance should be part of a new method on `ConnectorEnvironment`.

**Comment by maljac** [comment]
I used `@classmethod` since it s kind of a self referential factory pattern. I am not sure if thats the right way or if a plain method is sufficient.

**Comment by guewen** [comment]
@maljac yep, fine

**Comment by maljac** [comment]
Issue closed in PR #108
